### PR TITLE
Enhanced telemetry graphing capabilities

### PR DESCRIPTION
**Pull request by cazlo**
# What

The overall goal of this PR is to add a local graphing runtime which can be used to show some pretty telem graphs during mission operations.

The most likely impl is prometheus + grafana, but I may get spicy and do something like openc3 cosmos.

# Why

Live telem graphs of many concurrent signals within the control system can help to analyze and improve the control system's reliability.

# How

Generally this will likely involve the following tasks:
- [x] machine setup.
  - [x] python setup. added uv for dependency mgmt and everything seems to be fine with python 3.12
  - [x] install ksp (e.g. from steam), install ckan, use ckan to install krpc
  - [x] install craft files for maxerva, minerva
  - [x] put some links about machine setup in README.md
  - [x] verify machine setup OK by starting ksp in sandbox mode and launching a maxerva to orbit OK with `uv run maxerva1.py` 
- [x] init graphing infra
  - [x] docker compose for prometheus + grafana
  - [x] notes in readme for how to start/stop. 
  - [x] maybe a makefile for the scripts
- [x] basic dashboard. POC to demonstrate all the stuff is hooked up OK
  - [x] emit telem to prometheus
  - [x] graph telem in grafana
- [x] enhanced dashboard
  - [x] abstract telem stuff in some way and/or make it cross cutting somehow. want to do this in a way which reduced merge conflicts with the upcoming big changes to stuff like PID infra being deved in the 2024-12-07 stream
  - [x] ~~deep dive into the system and add telem that seems interesting~~ deferred for time
  - [x] ~~display the info done in the print like `Mode X Outputs:`~~ deferred for time
  - [x] add telem from ksp which is input into the system or is otherwise intersting like altitude
  - [x] complete refactor to add telem_viz hooks into avionics other than maxerva
- [x] podman support
  - [x] smokecheck on podman fedora box works ok
  - [x] for volume mounts use `:Z` instead of `:ro`
  - [x] makefile variable container provider
  - [x] docs (e.g. podman-compose dependency, podman-desktop recomendation)
- [x] DCX
  - [x] get latest craft file from discord
  - [x] smokecheck current state
  - [x] instrument dcx avionics with custom Gauges outputting gnc internals
  - [x] DCX specific dashboard
  - [x] instrument all the DCX variants
- [x] ~~(stretch goal) influx db instead of prometheus (probably as separate branch to see how much changes with new metric provider in current telem publish abstraction)~~ https://github.com/cazlo/ksp-krpc-rocket-avionics/pull/1
- [x] instrument performance and use instrumentation to find and fix performance issues. SLI is frame overruns

### CODE REVIEW

**Comment by cazlo** [comment]
Screenshot_20241208_120338

**Comment by cazlo** [comment]
10 hz graphs look good and have no noticable perf hit on my 32 core machine. tests on 780m mobile machine pending, may want to back it down to 1hz polling for laptop.

File: infra/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.12
+
+COPY . avionics
```
Comment (cazlo) [inline infra/Dockerfile]: should only be copying over here the uv control files like pyproject.toml and uv.lock, so when only python code changes we dont have to redownload dependencies when rebuilding the container image.

as is now the dependency list is small and downloads fast so didnt optimize from the get go

File: steering_logic.py [hunk 2]
```diff
@@ -3,25 +3,28 @@
...
+
 
 # Consists of functions for steering and controlling a generic launch vehicle.
 # - Pitch Maneuver
 # - Roll Program
 # - MaxQ throttle
 # - constant acceleration
 
-def roll_program(mission_params, telem, vessel, conn, sc):
+def roll_program(mission_params, telem, vessel, conn, sc, telem_viz: KSPTelemetry):
```
Comment (cazlo) [inline steering_logic.py:16]: to facilitate refactor, should set this with default to None, so it doesnt break all the other avionics and the refactor can be done in a more controlled manner, with same feedback applying to all other util fns changed in `launch_utils` and `steering_logic`

Comment (cazlo) [reply steering_logic.py:16]: encapsulated this in `enter_contrl_mode` fn

File: infra/Dockerfile (hunk 1 above)
Comment (cazlo) [reply infra/Dockerfile]: fixed

File: infra/grafana/dashboards/service.json [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+                    "MAX_Q",
+                    "THROTTLE_BUCKET_EXIT",
+                    "MECO_PREP",
+                    "MECO",
+                    "S2_INIT",
+                    "S2_LOOP1",
+                    "S2_LOOP2",
+                    "S2_LOOP3",
+                    "COAST",
+                    "ABORT"
```
Comment (cazlo) [inline infra/grafana/dashboards/service.json]: it's annoying all these enums are duplicated between the dashboard definition and the python code.

probably best long term solution would be to generate this list based on python enum state at docker build time.

in meantime remember to add control mode states both here and in the launch_utils control mode enum for proper display

**Comment by cazlo** [comment]
running the whole sim no problem on laptop sitting in power save mode (screen cap here is taking up ~1 CPU core for software rendering)

https://github.com/user-attachments/assets/18a29ea8-959f-42f4-8e78-7373872ac05f

**Comment by cazlo** [comment]
DCX GNC looking cool

Screenshot_20241210_152656

and looks even cooler when setting min step size to 10ms in graph settings:

Screenshot_20241210_160740

File: infra/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM python:3.12-alpine
+
+COPY .python-version pyproject.toml uv.lock avionics/
+
+WORKDIR avionics
+
+RUN pip install uv && uv sync
```
Comment (cazlo) [inline infra/Dockerfile]: lock to specific uv version to avoid code rot

File: launch_utils.py [hunk 5]
```diff
@@ -12,7 +57,7 @@ def initialize():
 
 
 def create_instance():
-    return krpc.connect(name='Launch')
+    return krpc.connect(name='Launch', address=os.environ.get('KRPC_ADDRESS', '127.0.0.1')) # todo add this to all instances like this
```
Comment (cazlo) [inline launch_utils.py:60]: might need to use `host.docker.internal` in windows/WSL. I don't have a windows box to test that behavior out on, but I left some commented code to make trying that override in the docker compose easy.

File: infra/Dockerfile (hunk 4 above)
Comment (cazlo) [reply infra/Dockerfile]: fixed

**Comment by cazlo** [comment]
Found a "quick win" to reducing overruns significantly (by about half). 

Here is a run which is representative of what the p95 of instrumented functions looks like without optimizations:

Screenshot_20241211_211112_function_timing_before_optimization

Note here the calculate_landing_burn_time p95 is well above 50hz or 30hz frame times. Not pictured here is the p50, which is also usually above the 50 or 30 hz frametime thresholds as well.

It mostly comes down to 2 optimizations. First is statically defining the flight_path_angle function instead of dynamically defining it every time calculate_landing_burn_time is called. As shown in the screenshot below, this change effectively removes the p95 spikes for flight_path_angle seen previously:

Screenshot_20241211_194447_get_flight_path_angle_optimization

The largest benefit was observed from only calculating the `g` and `engine_offset` values once instead of on each avionics compute frame tick. After this change overruns during mode 2 were reduced by about 50%, with calculate_landing_burn_time p95 largely staying within time budget. After change:

Screenshot_20241211_202834

Screenshot_20241211_212004

Screenshot_20241211_211905

Sample size is small but have not seen runs > 20m out of landing zone after change

**Comment by cazlo** [comment]
Biggest win for overruns/deterministic performance so far measured by refactoring from using blocking RPC calls for getting vehicle telem avionics needs to make decisions to using the stream API to get "latest" data.

This looks to be a big refactor so I split it into its own PR at https://github.com/cazlo/ksp-krpc-rocket-avionics/pull/2

tl;dr can run this at 50hz nearly without overruns (except in mode 2a) and at 40hz with only  a handful of overruns throughout

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,33 @@
+container_backend=podman
```
Comment (cazlo) [inline Makefile:1]: should probably default to docker since more ppl seem to use that.

also add some note about how this doesnt work as expected with rootless docker and network == host

File: dcx_flight_env_expansion copy.py [hunk 7]
```diff
@@ -214,31 +253,54 @@ def compute_heading_error(current_heading, target_heading):
 pitch_lpf = LPF(4,1,10.0)
 prev_dist = 0
 vel_sign = -1
+expected_frame_time = (1/dcx.CLOCK_RATE) * 1e9
+def dynamic_sleep(actual_frame_time: float):
```
Comment (cazlo) [inline dcx_flight_env_expansion copy.py:257]: should pull this into a util svc since it can be reused in most cyclic real time avionics for more determinism

File: infra/grafana/grafana.ini [hunk 8]
```diff
@@ -0,0 +1,1947 @@
...
+
+#################################### Dashboards ##################
+[dashboards]
+# Number dashboard versions to keep (per dashboard). Default: 20, Minimum: 1
+;versions_to_keep = 20
+
+# Minimum dashboard refresh interval. When set, this will restrict users to set the refresh interval of a dashboard lower than given interval. Per default this is 5 seconds.
+# The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
+################# non-default value for ksp-krpc-rocket-avionics below!
+min_refresh_interval = 100ms
```
Comment (cazlo) [inline infra/grafana/grafana.ini:448]: this is the only line changed from default config

File: telemetry.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            self.periapsis.set(telemetry.periapsis())
+
+            self.system_cpu_percent.set(psutil.cpu_percent())
+
+            time.sleep(1/self.publish_rate_hz)
+
+    # -------------------------------------------------------------
+    # Worker Thread Function
+    # -------------------------------------------------------------
+    def metric_publisher(self, queue: Queue):
```
Comment (cazlo) [inline telemetry.py:75]: for consistency with other threaded thing in this class, this fn should be internal and called by a fn like:

```suggestion

    def start_metrics_queue_thread(self, queue: Queue) -> bool:
        updater_thread = threading.Thread(target=self._metric_publisher_thread, args=(queue,), daemon=True)
        updater_thread.start()
        return True
    
      def _metric_publisher_thread(self, queue: Queue):
```

this would be nice for consumers b/c it would cut down on work of needing to start the thread

File: telemetry.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+                break
+
+            name = metric_data["name"]
+            type = metric_data["type"]
+            value = metric_data["value"]
+
+            if type == "gauge":
+                self.gauge_metrics[name].set(value)
+            else:
+                print(f"warning: unsupported metric type {type}")
```
Comment (cazlo) [inline telemetry.py:93]: would be nice to also support counts for the frame counters at least

dont see any use for supporting histograms

File: telemetry.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+    def register_histogram_metric(self, name: str, description: str):
+        custom_buckets = (.001, .005, .01, .02, .025, .033, .05, .075, .1, .25, .5, .75, 1.0, 2.5, 5.0, 7.5, 10.0, float("inf"))
+        self.histogram_metrics[name] = Histogram(name, description, buckets=custom_buckets)
+
+
+    def get_histogram_metric(self, name: str):
+        if name not in self.histogram_metrics:
+            print(f"WARNING: skipping publish of unknown histogram metric '{name}'. remember to register it first with register_histogram_metric!")
+            return
```
Comment (cazlo) [inline telemetry.py:138]: in current code state, if this case ever hit (or also the exception branch) it would still blow everything up b/c the histograms are basically only used for fn timers now.

probably better behavior is to still print the warning once to make it easy for dev to spot things like possible typo, but register the histogram here so the whole program execution doesnt blow up

File: telemetry.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+import threading
+from queue import Queue
+
+from prometheus_client import start_http_server, Gauge, Enum, Counter, Histogram
+import psutil
+
+from mission import Telemetry
+
+
+class KSPTelemetry:
```
Comment (cazlo) [inline telemetry.py:13]:
```suggestion
class TelemetryService:
```

File: telemetry.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+            if metric_data is None:
+                # This can be a signal to gracefully stop if you ever need it
+                break
+
+            name = metric_data["name"]
+            type = metric_data["type"]
+            value = metric_data["value"]
+
+            if type == "gauge":
+                self.gauge_metrics[name].set(value)
```
Comment (cazlo) [inline telemetry.py:91]: should probably use publish_gauge_metric for normalized error workflow (mostly for situation around unregistered metrics)

File: telemetry.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+from prometheus_client import start_http_server, Gauge, Enum, Counter, Histogram
+import psutil
+
+from mission import Telemetry
+
+
+class KSPTelemetry:
+    """
+    A class to handle publishing metrics periodically
```
Comment (cazlo) [inline telemetry.py:15]: given super positive results of POC of stream API at https://github.com/cazlo/ksp-krpc-rocket-avionics/pull/2

thinking to refactor this to abstract at least read telemetry calls to kRPC (automagically?) using stream API. also abstracting write calls to kRPC affords additional opportunity to telemeter output from the control system. some insights we can get from instrumenting control system output includes measurement of input lag, potential mistakes during control system dev, etc.

munging up these 2 concerns allows us to publish telem we use as input into the avionics system in a cross cutting way, with little/no effort needed in avionics scripts to setup the metrics system (i.e. could remove most or all of the `publish throttle telem after update`, `register metric for publish` stuff done during script init which is duplicated across avionics scripts in current iteration of this PR)

Comment (cazlo) [reply telemetry.py:15]: if this ends up getting done in this PR, make sure to add a diagram showing the threading model and how it is used to reduce impact of instrumentation on the code itself
